ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -74,995 +99,576 @@
+                                .times(min(total_penalties.times(UnsignedLong.valueOf(3L)), total_balance))
+                                .dividedBy(total_balance),
+                        validator.getEffective_balance()
+                                .dividedBy(UnsignedLong.valueOf(Constants.MIN_SLASHING_PENALTY_QUOTIENT)));
+        decrease_balance(state, index, penalty);
       }
-      index++;
     }
   }
 
   /**
-   * Process the exit queue. Note that this function mutates ``state``.
-   *
+   * v0.7.1
+   * https://github.com/ethereum/eth2.0-specs/blob/v0.7.1/specs/core/0_beacon-chain.md#final-updates
+   * Processes final updates
    * @param state
    */
-  public static void process_exit_queue(BeaconState state) {
-    // Sort in order of exit epoch, and validators that exit within the same epoch exit in order of
-    // validator index
-    List<Integer> sorted_indices =
-        IntStream.range(0, state.getValidator_registry().size())
-            .boxed()
-            .filter(i -> eligible(state, i))
-            .sorted(Comparator.comparing(i -> state.getValidator_registry().get(i).getExit_epoch()))
-            .collect(Collectors.toList());
+  public static void process_final_updates(BeaconState state) {
+    UnsignedLong current_epoch = get_current_epoch(state);
+    UnsignedLong next_epoch = current_epoch.plus(UnsignedLong.ONE);
 
-    int index = 0;
-    for (Integer dequeues : sorted_indices) {
-      if (dequeues >= Constants.MAX_EXIT_DEQUEUES_PER_EPOCH) {
-        break;
-      }
-      prepare_validator_for_withdrawal(state, index);
+    // Reset eth1 data votes
+    if (state.getSlot().plus(UnsignedLong.ONE).mod(UnsignedLong.valueOf(Constants.SLOTS_PER_ETH1_VOTING_PERIOD)).equals(UnsignedLong.ZERO)) {
+      state.setEth1_data_votes(new ArrayList<>());
     }
-  }
 
-  private static boolean eligible(BeaconState state, int index) {
-    Validator validator = state.getValidator_registry().get(index);
-    // Filter out dequeued validators
-    if (!validator.getWithdrawable_epoch().equals(FAR_FUTURE_EPOCH)) {
-      return false;
-    } else {
-      // Dequeue if the minimum amount of time has passed
-      return get_current_epoch(state)
-              .compareTo(
-                  validator
-                      .getExit_epoch()
-                      .plus(UnsignedLong.valueOf(Constants.MIN_VALIDATOR_WITHDRAWABILITY_DELAY)))
-          >= 0;
+    // Update effective balances with hysteresis
+    for (int index = 0; index < state.getValidator_registry().size(); index++) {
+      Validator validator = state.getValidator_registry().get(index);
+      UnsignedLong balance = state.getBalances().get(index);
+      long HALF_INCREMENT = Constants.EFFECTIVE_BALANCE_INCREMENT / 2;
+      if (balance.compareTo(validator.getEffective_balance()) < 0
+              || validator.getEffective_balance().plus(UnsignedLong.valueOf(3 * HALF_INCREMENT)).compareTo(balance) < 0) {
```

Comment:
I know this is the order of operations as given in the spec, but I have doubts whether the spec is right. Let's think about this one a little more.

Reply:
Agreed. Wanna ask this on gitter?